csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedInitialization.qll

```diff
@@ -0,0 +1,621 @@
+    TTranslatedExplicitElementInitialization, InitializationContext {
+  int elementIndex;
+
+  TranslatedExplicitElementInitialization() {
+    this = TTranslatedExplicitElementInitialization(initList, elementIndex)
+  }
+
+  override Instruction getTargetAddress() { result = getInstruction(getElementAddressTag()) }
+
+  override Type getTargetType() { result = getElementType() }
+
+  override Instruction getInstructionSuccessor(InstructionTag tag, EdgeKind kind) {
+    result = TranslatedElementInitialization.super.getInstructionSuccessor(tag, kind)
+    or
+    tag = getElementAddressTag() and
+    result = getInitialization().getFirstInstruction() and
+    kind instanceof GotoEdge
+  }
+
+  override Instruction getChildSuccessor(TranslatedElement child) {
+    child = getInitialization() and result = getParent().getChildSuccessor(this)
+  }
+
+  override TranslatedElement getChild(int id) { id = 0 and result = getInitialization() }
+
+  override int getElementIndex() { result = elementIndex }
+
+  TranslatedInitialization getInitialization() {
+    result = getTranslatedInitialization(initList.getElement(elementIndex))
+  }
+}
+
+/**
+ * Represents the IR translation of the initialization of a range of array
+ * elements without corresponding elements in the initializer list.
+ */
+class TranslatedElementValueInitialization extends TranslatedElementInitialization,
+  TTranslatedElementValueInitialization {
+  int elementIndex;
+
+  int elementCount;
+
+  TranslatedElementValueInitialization() {
+    this = TTranslatedElementValueInitialization(initList, elementIndex, elementCount)
+  }
+
+  override predicate hasInstruction(
+    Opcode opcode, InstructionTag tag, Type resultType, boolean isLValue
+  ) {
+    TranslatedElementInitialization.super.hasInstruction(opcode, tag, resultType, isLValue)
+    or
+    tag = getElementDefaultValueTag() and
+    opcode instanceof Opcode::Constant and
+    resultType = getDefaultValueType() and
+    isLValue = false
+    or
+    tag = getElementDefaultValueStoreTag() and
+    opcode instanceof Opcode::Store and
+    resultType = getDefaultValueType() and
+    isLValue = false
+  }
+
+  override Instruction getInstructionSuccessor(InstructionTag tag, EdgeKind kind) {
+    result = TranslatedElementInitialization.super.getInstructionSuccessor(tag, kind)
+    or
+    kind instanceof GotoEdge and
+    (
+      tag = getElementAddressTag() and
+      result = getInstruction(getElementDefaultValueTag())
+      or
+      tag = getElementDefaultValueTag() and
+      result = getInstruction(getElementDefaultValueStoreTag())
+      or
+      tag = getElementDefaultValueStoreTag() and
+      result = getParent().getChildSuccessor(this)
+    )
+  }
+
+  override string getInstructionConstantValue(InstructionTag tag) {
+    result = TranslatedElementInitialization.super.getInstructionConstantValue(tag)
+    or
+    tag = getElementDefaultValueTag() and
+    result = getZeroValue(getElementType())
```

Comment:
What about `null` elements?

Reply:
The `TranslatedElementValueInitialization` will not be used (I will probably delete it when I will refactor the whole `TranslatedInitialization.qll` file), since when an array is created and initialized we have 2 possibilities: either the size is specified, in which case the initializer list must be that size, or the size is not specified and we initialize as many elements as the initializer list has. Should I refactor this file for this PR?